### PR TITLE
Add mdg:validated-method and SimpleSchema

### PR DESCRIPTION
**Pull request by morloy**
We recently started typing our Meteor method using the definitions from this repository. Thanks a lot for the work, it’s already pretty awesome!
To go a step further, we also wanted to have flow check our method calls to the server. This PR adds typings for `mdg:validated-method` and `simpl-schema`.

Quick example to show how this works:
```
import { ValidatedMethod } from 'meteor/mdg:validated-method';
import SimpleSchema from 'simpl-schema';

export const getGreeting: Meteor$ValidatedMethod<{| name: string |}, string> = new ValidatedMethod({
  name: 'getGreeting',
  schema: new SimpleSchema({ name: String }).validator(),
  run({ name }) {
   return `Hello ${name}`;
  }
});

const greeting: string = getGreeting.call({ name: 'Bob' }); // works
const greeting: string = getGreeting.call(); // error
const n: number = getGreeting.call({ name: 'Bob' }); // error
```

Also in the function body of `run`, Flow will give you an error if in the above function you would try to write
```
run({ _id, name }) { … } // '_id' is not defined
```

### CODE REVIEW

**Comment by radekmie** [comment]
Hi @morloy. Glad you like it! Let's discuss it in parts. How about separating non-core packages? Maybe `meteor_mdg:validated-method_vx.x.x.js` -> `packages/mdg:validated-method_vx.x.x.js`? As for SimpleSchema, I'm against. It should be either in flow-typed (as it's an npm package) or in the package itself. It also means, that `Meteor$ValidatedMethod.schema` should be removed as it's not standard.

File: meteor_meteor_vx.x.x.js [hunk 1]
```diff
@@ -105,7 +107,11 @@ declare class Meteor$Meteor {
     retryTime?: number,
     reason?: string
   |};
-  subscribe(name: string, ...args: mixed[]): Meteor$SubscriptionHandle;
+
+  subscribe(name: string): Meteor$SubscriptionHandle;
+  // Define all subscription that need a parameter here, e. g.
+  // subscribe(name: 'posts', limit: number): Meteor$SubscriptionHandle;
```
Comment (radekmie) [inline meteor_meteor_vx.x.x.js]: I'd rather not change it, but the comment is really useful.

File: meteor_mongo_vx.x.x.js [hunk 2]
```diff
@@ -53,6 +55,13 @@ declare class Meteor$Mongo$Collection<BaseEntryT> {
       modifier: Meteor$Mongo$Modifier
     ) => boolean
   }): boolean;
+  attachSchema(
+    schema: SimpleSchema,
+    options?: $Shape<{
+      selector: number | string | Meteor$Mongo$ObjectId | Meteor$Mongo$Selector,
+      transform: boolean
+    }>
+  ): void;
```
Comment (radekmie) [inline meteor_mongo_vx.x.x.js]: It's also a non-standard method.

File: meteor_mdg:validated-method_vx.x.x.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    name: string,
+    validate?: Object => mixed,
+    run: T => ResultT
+  }): void;
+  call(args: T, fn?: (?Meteor$Error, ResultT) => mixed): ResultT;
+  name: string;
+  run(T, ...mixed[]): ResultT;
+  validate(Object): mixed;
+
+  schema: ?SimpleSchema;
```
Comment (radekmie) [inline meteor_mdg:validated-method_vx.x.x.js]: These are non-standard fields - let's remove it.

File: meteor_mongo_vx.x.x.js [hunk 4]
```diff
@@ -17,6 +17,8 @@ declare class Meteor$Mongo$BulkOpHandle {
   updateOne(modifier: Meteor$Mongo$Modifier): void;
 }
 
+declare type OptionalId<T> = {|...$Exact<T>, _id: ?string|};
```
Comment (radekmie) [inline meteor_mongo_vx.x.x.js]: What about `Meteor$Mongo$ObjectId`? I'd rather not separate this type. It's a useful one, but it pollutes types space.

File: meteor_mongo_vx.x.x.js [hunk 5]
```diff
@@ -188,8 +198,16 @@ declare class Meteor$Mongo$ObjectId {
 declare class Meteor$Mongo$RawCollection<T> {
   bulkWrite(pipeline: {}[]): Promise<{}>;
   deleteMany(selector: Meteor$Mongo$Selector): Promise<{}>;
+  distinct(field: string, query?: Meteor$Mongo$Selector): Promise<T[]>;
```
Comment (radekmie) [inline meteor_mongo_vx.x.x.js]: It's definitely not `T[]`. It should have own type parameter.

File: meteor_mongo_vx.x.x.js [hunk 6]
```diff
@@ -132,6 +141,7 @@ declare class Meteor$Mongo$Collection<BaseEntryT> {
     selector: number | string | Meteor$Mongo$ObjectId | Meteor$Mongo$Selector,
     callback: (?Error) => mixed
   ): void;
+  simpleSchema: () => SimpleSchema;
```
Comment (radekmie) [inline meteor_mongo_vx.x.x.js]: It's also a non-standard one.

File: meteor_mongo_vx.x.x.js [hunk 7]
```diff
@@ -188,8 +198,16 @@ declare class Meteor$Mongo$ObjectId {
...
   deleteMany(selector: Meteor$Mongo$Selector): Promise<{}>;
+  distinct(field: string, query?: Meteor$Mongo$Selector): Promise<T[]>;
   insertMany(docs: T[]): Promise<{}>;
   initializeUnorderedBulkOp(): Meteor$Mongo$BulkOp;
+  createIndex(
+    keys: {[string]: 1 | -1},
+    options?: {
+      name?: string,
+      unique?: boolean
+    }
```
Comment (radekmie) [inline meteor_mongo_vx.x.x.js]: How about making it exact?

File: meteor_webapp_vx.x.x.js [hunk 8]
```diff
@@ -52,6 +53,7 @@ declare type Meteor$WebApp$Request = {
 };
 
 declare class Meteor$WebAppInternals {
+  enableSubresourceIntegrity(use_credentials?: boolean): void;
```
Comment (radekmie) [inline meteor_webapp_vx.x.x.js]: Camel case, please.

**Review by radekmie — CHANGES_REQUESTED** [review]

File: meteor_mongo_vx.x.x.js (hunk 2 above)
Comment (morloy) [reply meteor_mongo_vx.x.x.js]: It’s standard in `aldeed:collection2`. Do you have any idea about how to differentiate standard `meteor/mongo` and `aldeed:collection2` in Flow?

Comment (radekmie) [reply meteor_mongo_vx.x.x.js]: That's actually a long-standing issue in Flow: facebook/flow#396. For now, you can either override it locally or create a custom `Meteor` type with the intersection operator:

```js
type Collection<T> = Meteor$Mongo$Collection<T> & {attachSchema(...): ...};
```

**Comment by radekmie** [comment]
Ping @morloy.

File: meteor_mongo_vx.x.x.js (hunk 2 above)
Comment (morloy) [reply meteor_mongo_vx.x.x.js]: The problem here is that `new Mongo()` still returns a `Meteor$Mongo$Collection`, so I’m unable to write something like
```
const Posts: Collection<PostProps> = new Mongo('posts');
```

Is there a way to re-declare the `declare module 'meteor/mongo'` part?

Comment (radekmie) [reply meteor_mongo_vx.x.x.js]: Nothing I know about. If you declare a new type (anyhow), you can cast (or force cast) your collections to have this type.

**Comment by morloy** [comment]
Separated `meteor/aldeed:collection2` from `meteor/mongo` and added some hints to the Readme. Hope that’s fine now.

File: README.md [hunk 9]
```diff
@@ -1,3 +1,26 @@
...
+module.name_mapper='^\/\(.*\)$' -> '<PROJECT_ROOT>/\1'
+module.name_mapper='^meteor\/\(.*\):\(.*\)$' -> '<PROJECT_ROOT>/.meteor/local/build/programs/server/packages/\1_\2'
+module.name_mapper='^meteor\/\(.*\):\(.*\)$' -> '<PROJECT_ROOT>/.meteor/local/build/programs/web.browser/packages/\1_\2'
+module.name_mapper='^meteor\/\(.*\)$' -> '<PROJECT_ROOT>/.meteor/local/build/programs/server/packages/\1'
+module.name_mapper='^meteor\/\(.*\)$' -> '<PROJECT_ROOT>/.meteor/local/build/programs/web.browser/packages/\1'
+```
+
+#### aldeed:collection2
+1. In `meteor_mongo_vx.x.x.js` remove the part saying `declare module 'meteor/mongo' {…}`
+2. In `meteor_aldeed:collection2_vx.x.x.js` replace `meteor/aldeed:collection2` with `meteor/mongo`.
```
Comment (radekmie) [inline README.md]: It's neither recommended nor valid. For example, `MongoInternals` are not present there. I'd rather suggest casting it manually.

File: meteor_mongo_vx.x.x.js (hunk 6 above)
Comment (radekmie) [reply meteor_mongo_vx.x.x.js]: This one is still not addressed.

File: meteor_mongo_vx.x.x.js [hunk 10]
```diff
@@ -17,6 +17,11 @@ declare class Meteor$Mongo$BulkOpHandle {
   updateOne(modifier: Meteor$Mongo$Modifier): void;
 }
 
+declare type OptionalId<T> = {|
+  ...$Exact<T>,
+  _id: null | string | Meteor$Mongo$ObjectId
```
Comment (radekmie) [inline meteor_mongo_vx.x.x.js]: It'd be better to use `$PropertyType<T, '_id'>`, to make sure it's the same type.

File: meteor_mongo_vx.x.x.js [hunk 11]
```diff
@@ -188,8 +194,16 @@ declare class Meteor$Mongo$ObjectId {
 declare class Meteor$Mongo$RawCollection<T> {
   bulkWrite(pipeline: {}[]): Promise<{}>;
   deleteMany(selector: Meteor$Mongo$Selector): Promise<{}>;
+  distinct(field: string, query?: Meteor$Mongo$Selector): Promise<mixed[]>;
```
Comment (radekmie) [inline meteor_mongo_vx.x.x.js]: Generic type instead of `mixed` is definitely better here.

File: packages/meteor_aldeed:collection2_vx.x.x.js [hunk 12]
```diff
@@ -0,0 +1,22 @@
+declare class Aldeed$Mongo$Collection2<T> extends Meteor$Mongo$Collection<T> {
```
Comment (radekmie) [inline packages/meteor_aldeed:collection2_vx.x.x.js]: Why `$Mongo`? It should be `Aldeed$Collection2` or `Aldeed$Collection2$Collection`.

File: packages/meteor_aldeed:collection2_vx.x.x.js [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+  simpleSchema: () => SimpleSchema;
+}
+
+declare class Aldeed$Mongo {
+  Collection: typeof Aldeed$Mongo$Collection2;
+  Cursor: typeof Meteor$Mongo$Cursor;
+}
+
+declare module 'meteor/aldeed:collection2' {
+  declare module.exports: {
```
Comment (radekmie) [inline packages/meteor_aldeed:collection2_vx.x.x.js]: It's not true, that this module exports such object. Either not define it at all, or define it properly.

File: packages/simpl-schema_vx.x.x.js [hunk 14]
```diff
@@ -0,0 +1,43 @@
+declare type SimpleSchema$ValidationError = {
```
Comment (radekmie) [inline packages/simpl-schema_vx.x.x.js]: Again, it's a npm package, so it should be submitted directly to [flow-typed](https://github.com/flow-typed/flow-typed).

**Review by radekmie — CHANGES_REQUESTED** [review]
We are not there (yet).

File: README.md (hunk 9 above)
Comment (morloy) [reply README.md]: True, changed that to forced type casting.

File: meteor_mongo_vx.x.x.js (hunk 11 above)
Comment (morloy) [reply meteor_mongo_vx.x.x.js]: Not sure I get that. You mean a template type? Using which template parameter?

File: packages/simpl-schema_vx.x.x.js (hunk 14 above)
Comment (morloy) [reply packages/simpl-schema_vx.x.x.js]: For ease of maintenance, I would really prefer to keep it in here. I’m not sure if anybody out there is using `simpl-schema` without Meteor, so it’s nice to have everything related to Meteor at one place.

File: meteor_mongo_vx.x.x.js (hunk 11 above)
Comment (radekmie) [reply meteor_mongo_vx.x.x.js]: Yep, I meant `distinct<X>(field: string, query?: Meteor$Mongo$Selector): Promise<X[]>;`. I don't have a name though.
`

File: packages/simpl-schema_vx.x.x.js (hunk 14 above)
Comment (radekmie) [reply packages/simpl-schema_vx.x.x.js]: I'm one of such people :)

Comment (morloy) [reply packages/simpl-schema_vx.x.x.js]: All right, so removing `simpl-schema` from this PR and submitting to `flow-typed` instead would be the last step left?

Comment (radekmie) [reply packages/simpl-schema_vx.x.x.js]: Yes, sorry for being so picky!

Comment (morloy) [reply packages/simpl-schema_vx.x.x.js]: Removed `simpl-schema`.

**Review by radekmie — APPROVED** [review]

**Pull request merged**